feat(arithmetic): Adding more column support to count_if

wmak opened the pull request:
- This adds duration column support to count_if along with general tags support
  - eg. `count_if(transaction.duration, less, 300)` or `count_if(http.status, equals 400)`
- This also adds count_if to arithmetic
- Not added to the frontend yet

wmak committed:
feat(arithmetic): Adding more column support to count_if

- This adds duration column support to count_if along with general tags
  support
- This also adds count_if to arithmetic
- Not added to the frontend yet

shruthilayaj commented on tests/snuba/api/endpoints/test_organization_events_v2.py:
```diff
@@ -4057,3 +4057,33 @@ def test_equation_field_limit(self):
 
         assert response.status_code == 400
+
+    def test_count_if(self):
```
Should we add a test case with a refining filter like "count_if(transaction.duration, less, 150): >2" or something?

shruthilayaj commented on src/sentry/search/events/fields.py:
```diff
@@ -1791,22 +1821,28 @@ def is_accessible(self, acl=None):
+            calculated_args=[
+                {
+                    "name": "typed_value",
+                    "fn": calculate_count_if_value,
```
nit: can we name this function to something like "normalize_count_if_value" or "type_count_if_value" since it isn't actually calculating anything

shruthilayaj commented on src/sentry/search/events/fields.py:
```diff
@@ -259,6 +260,33 @@ def team_key_transaction_expression(organization_id, team_ids, project_ids):
+            raise InvalidSearchQuery(f"{value} is not a valid value to compare with {column}")
+
+    # TODO: not supporting field aliases or arrays yet
+    elif column in FIELD_ALIASES or column in ARRAY_FIELDS:
+        raise InvalidSearchQuery(f"{column} not supported by count_if")
+    # At this point only string or tag columns are left
+    elif condition not in ["equals", "notEquals"]:
+        raise InvalidSearchQuery(f"{condition} not compatible with {column}")
```
since this column validation has nothing to do with the typing of the value, can this be validated in one of the subclasses of the FunctionArgs class? It we do this kind of validation in the `normalize` method there generally. We  may have to introduce a countIf column type or something though

wmak replied:
So ended up with this cause the normalize doesn't have the context of the other variables, and I think its better if normalize doesn't need that. So calculated args seemed like a better fit to me

Zylphrex commented on src/sentry/search/events/fields.py:
```diff
@@ -259,6 +260,33 @@ def team_key_transaction_expression(organization_id, team_ids, project_ids):
+    column = args["column"]
+    condition = args["condition"]
+    value = args["value"]
+    if (
+        column == "transaction.duration"
+        or is_duration_measurement(column)
+        or is_span_op_breakdown(column)
+    ):
+        try:
+            value = float(value.strip("'"))
+        except Exception:
+            raise InvalidSearchQuery(f"{value} is not a valid value to compare with {column}")
```
What about numeric fields that aren't durations? e.g. `measurements.cls`

Zylphrex commented on src/sentry/search/events/fields.py:
```diff
@@ -1791,22 +1821,28 @@ def is_accessible(self, acl=None):
             "count_if",
             required_args=[
-                ColumnNoLookup("column", allowed_columns=["event.type", "http.status_code"]),
+                FunctionArg("column"),
```
Why is this a `FunctionArg` instead of `ColumnNoLookup` now? This removes the validation that it's a real column right? Is this to support tags as well?

wmak replied:
Exactly, this is cause I want to support tags like `http.status_code` i'll add a comment :+1:

Zylphrex commented on tests/sentry/search/events/test_fields.py:
```diff
@@ -1398,3 +1398,71 @@ def test_redundant_grouping_errors(self):
+        ]
+
+    def test_invalid_count_if_fields(self):
+        with self.assertRaises(InvalidSearchQuery):
+            resolve_field_list(["count_if(project, equals, sentry)"], eventstore.Filter())
+
+        with self.assertRaises(InvalidSearchQuery):
+            resolve_field_list(["count_if(stack.function, equals, test)"], eventstore.Filter())
```
Can we add an assertion to make it obvious why these are invalid?

wmak committed:
ref: Addressing PR comments

- Addingg test that includes filtering on ocunt_if
- Renaming to normalize_count_if_value
- using is_measurement instead of is_duration_measurement
- Adding comment exlpaining why FunctionArg
- Adding asserts on error messages
  - Also adding more tests for other potential errors

Zylphrex commented on src/sentry/search/events/fields.py:
```diff
@@ -954,8 +980,8 @@ def normalize(self, value, params):
 class ColumnNoLookup(Column):
-    def __init__(self, name, allowed_columns=None):
-        super().__init__(name, allowed_columns=allowed_columns)
+    def __init__(self, name, **kwargs):
+        super().__init__(name, **kwargs)
```
Is this some intermediate change that made it into the PR? I don't see it being used anywhere.

wmak replied:
just leftover changes cause I was originally going to extend `allowed_columns` until I realized that doing so wouldn't work with tags. 

gonna just remove this init entirely now I'm reading it again since `ColumnNoLookup` doesn't do anything here

Zylphrex reviewed (approved):
Just a comment about the unused changes to `ColumnNoLookup` but otherwise LGTM

wmak committed:
ref: cleanup
